Fix-enh bcf v2

sem-geologist opened the pull request:
This PR fixes bugs which were found with loading version 2 bcf #1502.
The first issue were previously some statically getting the xml nodes from the bcf header part,
this patch changed that to use xpath which allows getting the failed nodes independently from the order.
The second issue is the hypercube loading failing (python pure parser) or segfaulting (cython implementation), it appeared that with previously tested files flag 0 never were met and were not implemented, causing the failure (previously only 1, 2, 3 flags were observed)
the 0 flag marked pixels are packed in 16 bit. So in the end it can be said that flag 0 - 16bit packed impulses, 1 - 12bit packed impulses, 2, 3+? - instructively packed arrays.

Finally, v2 bcf contains mapping overview, which would be handy to return in the signal list. The Overview image contains rectangle information of mapping area which allows to reconstruct the marker.

Tasks:

- [x] fix the xml parser
- [x] fix the datacube pure python reading implementation
- [x] fix the datacube cython reading implementation
- [x] add loading of overview image (v2 only)
- [x] add rectangular marker to overview image marking the region of hypermap (v2 only)
- [x] add the docstring of bcf-only parameters to main hspy load function
- [x] do not try reading overview if there is no overview (on v2, thanks to @ericpre )

sem-geologist committed:
Merge pull request #19 from hyperspy/RELEASE_next_patch

update with upstream

sem-geologist committed:
fix: refering to xml nodes
using xpath instead of index which
life showed to be in different order
- py_parse_hypermap implemented
flag 0, which means that pixel
is packed in 16bit impulses

sem-geologist committed:
fix: unbcf_fast.pyx
add the 16bit impuls unpacking

sem-geologist commented:
@francisco-dlp , I think WIP flag can be removed, This should work

sem-geologist commented:
damn... I see I again pushed that to wrong branch...

sem-geologist commented:
This is ready for review

francisco-dlp commented:
I get a segfault when loading the dataset.

sem-geologist commented:
did you removed c files and recythonized? and build_ext --inplace ?

francisco-dlp commented:
Yes.

francisco-dlp commented:
Before merging this we also need a smaller file for the test. I'll see if somebody here can provide it.

sem-geologist commented:
best would be if the image itself (overview) would not exceed 100pixels, in that way the file will be really small

sem-geologist commented:
I think setup.py and handling recythonization is a bit broken, I see that it does not remove old c file, I also was getting segfault, until I manually removed C file in io_plugins. I will have to deal with this problem in separate PR. @francisco-dlp  if you remove c file manualy and do recythonize, does it still segfaults?

francisco-dlp commented:
You may be right. After deleting it manually, recythonizing and building in place now it doesn't segfault, but it doesn't use the cython version either :(

sem-geologist commented:
that is weird... are you sure there is no *.so files produced?

francisco-dlp commented:
The so file is there, but somehow it fails to import it.

francisco-dlp commented:
This is the error:
```
ImportError: /home/fjd29/Python/hyperspy3/hyperspy/io_plugins/unbcf_fast.cpython-35m-x86_64-linux-gnu.so: undefined symbol: PyFPE_jbuf
```

But I think that this is might be a particular problem of my installation. So I would say that this is good to merge, but we have to wait for a test file.

sem-geologist commented:
are you sure you use hyperspy in development mode? maybe it pushes the library to some global folder, so it would have no extension there. only built in place. Try building without `--inplace`, or check if hyperspy when imported is using the same patch where is the source.

sem-geologist commented:
https://github.com/scikit-learn/scikit-learn/issues/7542

and this is why I don't use (and am against) conda on linux

sem-geologist committed:
add: overview figure for bcf v2

sem-geologist commented:
@francisco-dlp , to long to wait for those images, so I added the overview image loading and rectangular marker :) to this PR. Now I need to wait for small file of bcf v2 to add it to test.

sem-geologist committed:
fix: guessing instrument
(TEM or SEM)

sem-geologist committed:
small cleanup

sem-geologist commented:
Good this still not landed to the main. There were some additional bugs found, 1.st the cython implementation is wrong and gives some gibberish for 16bit packed pixels.
2nd. I feel urge to share excerpt from the bcf xml part (pay attention to the 3rd line):
```xml
<ClassInstance Type="TRTSpectrumRegion" Name="Au">
    <TRTBase>
       <Description>Au-MAB</Description>
    </TRTBase>
    <Element>79</Element>
    <Line>MAB</Line>
    <Energy>2.149462777</Energy>
    <Width>1.644548913E-1</Width>
    <ProcWidth>8.7E-1</ProcWidth>
    <Color>16711680</Color>
  </ClassInstance>
```
this `Au-MAB` bugger hit me unexpected. What a shitty, lazy Bruker'ish way to save the space.

How I should fix it?

1. because it is the region of close lines, I could chose to return only one: I could check if line representing string length is >2 then crop it to 2 leaving i.e. MA.
2. check if it is >2 length and generate two lines Ma and Mb.

(to implement 1 is very easy, 2 is more complicated and will make things more cluttered)

sem-geologist committed:
clean-doc: updated the struct reading
1. pixel size actually can be longer than 16bit address
and empty struct taken for unknow is part of 32bit address
2. updated struct comment description

sem-geologist committed:
fix: Bruker'ish mind shortcut
where few close lines were described as one
causing hyperspy not to recognise the parsed line:
i.e. Au-MAB converted to hyperspy Au_Mab;
fix cutsoff anything above 2 strings
so it is returned as Au_Ma

sem-geologist committed:
fix: fix the 16bit pulse parsing
previously left byte was accidently set
to be moved by 8 bit right, causing
the number to be discarded.

sem-geologist commented:
Everything is fixed now.

@francisco-dlp , I should mention that I was really skeptical about leaving pure python bcf binary parser function as it is ugly and long. But this issue just proved how wrong I was, it was much easier to find the problem and fix it in python parser, and to spot and troubleshoot the new problem introduced into cython code.

sem-geologist committed:
cleanup: remove bzip2 from supported
compression list  as it is not
used by bruker, neither it is implemented
in the bcf library.

sem-geologist committed:
add: bcf 4x3 pixel hypermap with 32bit size values

sem-geologist committed:
add: docstring to main load hspy function
describing optional bcf only options

sem-geologist committed:
add: v2 ncf file to tests

sem-geologist committed:
update test to include v2 bcf

sem-geologist committed:
typo fix

sem-geologist commented:
@francisco-dlp , it is ready for review and merge, v2 bcf was added for tests, failures of travis and appveyor is due to `interpolate` changes in scipy, there are no traces in the travis and appveyor logs of  any bcf failures. The coverage of `io_plugins/bcf.py` with two new bcf samples got up to 93% from 89%.

francisco-dlp commented on hyperspy/io.py:
```diff
@@ -112,6 +112,23 @@ def load(filenames=None,
         overview.
+        
+    downsample : int (1–4095)
+        For Bruker bcf files, if set to integer (>=2) (default 1)
```
The language might need some improvement. Would any native speaker like to help (e.g. @tjof2)?

If not, I can help making it more spanglish...

sem-geologist replied:
well... I tried it to sound in the same style as doc "For SEMPER unf- and EMD...." line above

ericpre committed:
Esprit 2 bcf files: don't try to read overview images if there is no images. It happen for files converted from FEI emd to bcf format using a FEI script.

sem-geologist committed:
Merge pull request #20 from ericpre/open_bcf2_with_missing_overview_image

Don't read overview images when there is no images

sem-geologist committed:
cleanup: some code cleanup

sem-geologist committed:
fix-update: original_metadata json test file

sem-geologist committed:
prevent lanscape from complaining about lines
where nothing can be done.

to266 commented:
LGTM. I'd merge if not the @francisco-dlp worry about English

sem-geologist commented:
It is written in the same style as SEMPER unf- and EMD part in the same docstring, I guess it can be beautified later?

francisco-dlp commented:
Here is a suggestion:

```
downsample : int (1–4095)
        For Bruker bcf files, if set to integer (>=2) (default 1)
        bcf is parsed into down-sampled size array by given integer factor,
        multiple values from original bcf pixels are summed forming downsampled
        pixel. This allows to improve signal and conserve the memory with the
        cost of lower resolution.
    cutoff_at_kV : {None, int, float}
       For Bruker bcf files, if set to numerical (default is None)
       bcf is parsed into array with depth cutoff at coresponding given energy.
       This allows to conserve the memory, with cutting-off unused spectra's
       tail, or force enlargement of the spectra size.
    select_type: {'spectrum', 'image', None}
       For Bruker bcf files, if one of 'spectrum' or 'image' (default is None)
       the loader returns either only hypermap or only SEM/TEM electron images.
```

francisco-dlp commented:
By the way, wouldn't ``bin`` be a better name for the parameter as it is binning, not downsampling, what it does? Also, the docstring does not clarify which axes are binned (the spatial axes only?). Also, is there any constraint of the factor (i.e. it must be an integer divisor or the spatial axes size?). And, why not adding the option to bin each axis independently (including the spectral axis) using the same syntax as in ``BaseSignal.rebin``?

sem-geologist commented:
@francisco-dlp yes it is spatial axes only, why you would want to downsample x and y independently?
There is no implementation in downsampling/binning of signal axe, as that would bring significant penalty in reading speed. Downsample as implemented comes with no cost as it uses the technical nuance of way the bcf is read -- pixel by pixel. Additionally it saves the memory, as pixels are filled into smaller size of array without creating the full size array. It does something like `skimage.measure.block_reduce()` with `sum` as the function, just differently than that it costs no memory and speed. It is useful when bcf mapping is gathered with high spatial resolution (which max is 4096x4096 pixels x >=1024 channels x >=8bit) where a lot of machines with less than 32GB memory would have trouble of opening such file.
Also it is useful for simultanious aquisition of high spatial resolution images (SEI, BSE, etc...) and hypermap (which at that resolution is noisy, but when downsampled is good enough).

I dont know if binning is better term than downsampling?

francisco-dlp commented:
That makes sense. As I understanding it, binning involves summing the several pixels into a single one while sownsampling involves selecting 1 of every n pixels. Both results in a dataset of the same size but through different methods and outcomes.

What about my question regarding the binning factor?

Regarding the binning axes, the main point was that it is not crystal clear (although it can be inferred) that it only applies to the spatial axes.

By the way, I've just noticed that I've pasted the wrong docstring.

francisco-dlp commented:
This is the suggested spanglish version:

```python

    bin : int (1–4095)
        For Bruker bcf files, if set to integer (>=2) (default 1)
        the spatial axes are binned by the given integer factor.  For
        example, if `bin` is 3, 3 consecutive pixels from the x-axis in the file are
        summed to obtain each x-axis pixel in the loaded data. Binning enhances the
        signal to noise ratio and decreases the size of the data in memory at
        the expense of spatial resolution.
    cutoff_at_kV : {None, int, float}
       For Bruker bcf files, if a number (default is None)
       the spectral data is cropped to the given `cutoff_at_kV` value in kV
       if below the axis upper limit, of padded up to this value otherwise.
       This is useful to reduce the size of the data in memory when the
       spectral range above a given value is not required.
    select_type: {'spectrum', 'image', None}
       For Bruker bcf files, if 'spectrum'/'image' (default is None)
       the load only multi-dimensional spectra/images.
```

francisco-dlp commented:
By the way, shouldn't it be keV instead of kV?

sem-geologist commented:
yes it should be `keV`,  I will need to change it in all the files (same as `downsample` to `bin`.
The factor constrain is > = 1. Anything more than size of spatial axes size will give the same effect as with max spatial size so it is even not checked as the only thing it would  bring the reading speed penalty. So where is practically no upper bound. Nevertheless it is suggested `int (1-4096)` as those are maximum pixel values which can be set in Bruker Esprit.

francisco-dlp commented:
So, what happens when it is not a divisor of the axis size e.g. binning 3 when the size is 10?

sem-geologist commented:
and btw, it is both, x and y axes binned:
with `bin=3`: (or >3)
[[3, 3, 3],[3,3,3],[3,3,3]] would be binned to [[27]]
with `bin=2`:
it would be binned to [[12,6],[6,3]];
nothing is done with the border, it is left for user to decide.

francisco-dlp commented:
@sem-geologist, we're thinking on releasing soon. Would you have time to finish this (it's almost ready except for some minor things and the merge conflicts).

idahj committed:
TYPO: fix typos in eels user guide

francisco-dlp committed:
Fixes #1625

francisco-dlp committed:
Merge pull request #1635 from francisco-dlp/FIX_remove_pl_background

Fix remove Power Law background background

idahj committed:
TYPO: Fix typos in some files in user guide

francisco-dlp committed:
Fix blockfile regression

francisco-dlp committed:
Merge pull request #1636 from idahj/TYPO_eels_user_guide

Fix typos in user guide

francisco-dlp committed:
Explain that RGB is only for plotting

francisco-dlp committed:
Merge pull request #1638 from francisco-dlp/DOC_RGB_is_for_plotting

Explain that RGB is only for plotting in the User Guide

francisco-dlp committed:
Merge pull request #1637 from francisco-dlp/FIX_blockfile_reader

Fix blockfile regression

francisco-dlp committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into fix_bcf_v2

francisco-dlp committed:
Automatic style corrections courtesy of autopep8

francisco-dlp commented:
I've fixed the conflicts so merging.
